File: nestjs-practica05/src/products/mappers/product.mapper.ts
import { CreateProductDto } from '../dtos/create-product.dto';
import { ProductEntity } from '../entities/product.entity';
import { ProductModel } from '../models/product.model';
import { ProductResponseDto } from '../dtos/product-response.dto';

export class ProductMapper {
  static toModelFromDto(dto: CreateProductDto): ProductModel {
    const model = new ProductModel();
    model.name = dto.name;
    model.price = dto.price;
    model.stock = dto.stock;
    return model;
  }

  static toModelFromEntity(entity: ProductEntity): ProductModel {
    const model = new ProductModel();
    model.id = entity.id;
    model.name = entity.name;
    model.price = Number(entity.price);
    model.stock = entity.stock;
    model.createdAt = entity.createdAt;
    model.updatedAt = entity.updatedAt;
    model.deleted = entity.deleted;
    return model;
  }

  static toEntityFromModel(model: ProductModel): ProductEntity {
    const entity = new ProductEntity();
    entity.name = model.name;
    entity.price = model.price;
    entity.stock = model.stock;
    return entity;
  }

  static toResponse(model: ProductModel): ProductResponseDto {
    const response = new ProductResponseDto();
    response.id = model.id;
    response.name = model.name;
    response.price = model.price;
    response.stock = model.stock;
    return response;
  }
}

File: src/main/java/ec/edu/ups/icc/fundamentos01/products/mappers/ProductMapper.java
package ec.edu.ups.icc.fundamentos01.products.mappers;

import ec.edu.ups.icc.fundamentos01.categories.mappers.CategoryMapper;
import ec.edu.ups.icc.fundamentos01.products.dto.CreateProductDto;
import ec.edu.ups.icc.fundamentos01.products.dto.ProductResponseDto;
import ec.edu.ups.icc.fundamentos01.products.entities.ProductEntity;
import ec.edu.ups.icc.fundamentos01.products.models.ProductModel;

import java.time.LocalDateTime;

public class ProductMapper {

    public static ProductEntity toEntity(CreateProductDto dto) {
        ProductEntity entity = new ProductEntity();
        entity.setName(dto.getName());
        entity.setPrice(dto.getPrice());
        entity.setStock(dto.getStock());
        entity.setCreatedAt(LocalDateTime.now());
        return entity;
    }

    public static ProductModel toModel(CreateProductDto dto) {
        ProductModel model = new ProductModel();
        model.setName(dto.getName());
        model.setPrice(dto.getPrice());
        model.setStock(dto.getStock());
        model.setCreatedAt(LocalDateTime.now());
        return model;
    }

    public static ProductResponseDto toResponse(ProductModel model) {
        ProductResponseDto response = new ProductResponseDto();
        response.setId(model.getId());
        response.setName(model.getName());
        response.setPrice(model.getPrice());
        response.setStock(model.getStock());
        return response;
    }

    public static ProductResponseDto toResponse(ProductEntity entity) {
        ProductResponseDto response = new ProductResponseDto();
        response.setId(entity.getId());
        response.setName(entity.getName());
        response.setPrice(entity.getPrice());
        response.setStock(entity.getStock());
        if (entity.getCategories() != null) {
            response.setCategories(entity.getCategories()
                    .stream()
                    .map(CategoryMapper::toResponse)
                    .toList());
        }
        return response;
    }
}
